pkg/controllers/workflow/controller_test.go

```diff
@@ -0,0 +1,709 @@
+							DHCP: &v1alpha1.DHCP{
+								Arch:     "x86_64",
+								Hostname: "sm01",
+								IP: &v1alpha1.IP{
+									Address: "172.16.10.100",
+									Gateway: "172.16.10.1",
+									Netmask: "255.255.255.0",
+								},
+								LeaseTime:   86400,
+								MAC:         "3c:ec:ef:4c:4f:54",
+								NameServers: []string{},
+								UEFI:        true,
+							},
+						},
+					},
+				},
+			},
+			req: reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      "debian",
+					Namespace: "default",
+				},
+			},
+			want: reconcile.Result{},
+			wantWflow: &v1alpha1.Workflow{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Workflow",
+					APIVersion: "tinkerbell.org/v1alpha1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "1000",
+					Name:            "debian",
+					Namespace:       "default",
+				},
+				Spec: v1alpha1.WorkflowSpec{
+					TemplateRef: "debian",
+					HardwareMap: map[string]string{
+						"device_1": "3c:ec:ef:4c:4f:54",
+					},
+				},
+				Status: v1alpha1.WorkflowStatus{
+					State:         "STATE_TIMEOUT",
+					GlobalTimeout: 600,
+					Tasks: []v1alpha1.Task{
+						{
+							Name:       "os-installation",
+							WorkerAddr: "3c:ec:ef:4c:4f:54",
+							Volumes: []string{
+								"/dev:/dev",
+								"/dev/console:/dev/console",
+								"/lib/firmware:/lib/firmware:ro",
+							},
+							Actions: []v1alpha1.Action{
+								{
+									Name:    "stream-debian-image",
+									Image:   "quay.io/tinkerbell-actions/image2disk:v1.0.0",
+									Timeout: 60,
+									Environment: map[string]string{
+										"COMPRESSED": "true",
+										"DEST_DISK":  "/dev/nvme0n1",
+										"IMG_URL":    "http://10.1.1.11:8080/debian-10-openstack-amd64.raw.gz",
+									},
+									Status:    "STATE_TIMEOUT",
+									StartedAt: TestTime.MetaV1BeforeSec(601),
+									Seconds:   601,
+									Message:   "Action timed out",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantErr: nil,
+		},
+	}
+
+	for _, tc := range cases {
+		kc := GetFakeClientBuilder()
+		if tc.seedHardware != nil {
+			kc = kc.WithObjects(tc.seedHardware)
+		}
+		if tc.seedTemplate != nil {
+			kc = kc.WithObjects(tc.seedTemplate)
+		}
+		if tc.seedWorkflow != nil {
+			kc = kc.WithObjects(tc.seedWorkflow)
+		}
+		controller := &Controller{
+			kubeClient: kc.Build(),
+			nowFunc:    TestTime.Now,
+		}
+
+		t.Run(tc.name, func(t *testing.T) {
+			got, gotErr := controller.Reconcile(context.Background(), tc.req)
+			if gotErr != nil {
+				if tc.wantErr == nil {
+					t.Errorf(`Got unexpected error: %v"`, gotErr)
+				} else if gotErr.Error() != tc.wantErr.Error() {
+					t.Errorf(`Got unexpected error: got "%v" wanted "%v"`, gotErr, tc.wantErr)
+				}
+				return
+			}
+			if gotErr == nil && tc.wantErr != nil {
+				t.Errorf("Missing expected error: %v", tc.wantErr)
+				return
+			}
+			if tc.want != got {
+				t.Errorf("Got unexpected result. Wanted %v, got %v", tc.want, got)
```

Comment:
why no return? Also, why not just `t.FailNow()` which if I read the docs right will stop this test but continue others.

Reply:
Yep, works: https://go.dev/play/p/m3xCy3B-onA

Reply:
We also want to compare the mutated object, not just the controller response